### DISCUSSION TITLE
Ошибка с семафорами, поезд поворачивает на красный сегмент при существовании зелёного

### DISCUSSION THREAD

**Discussion by Ascron** [voxeltycoon/issues#736] [Troubleshooting]
**Describe the bug**
Встречный разъезд для двух поездов, поезд постоянно едет постоять на красный сигнал, хотя рядом есть зелёный.

**Steps to reproduce**
https://cdn.discordapp.com/attachments/129503130894336000/840718510816231494/2021-05-09_01-29-37.mp4

**Build version**
0.85.2.1 steam

**Platform**
Выпуск	Windows 10 Домашняя
Версия	20H2
Дата установки	‎13.‎07.‎2020
Сборка ОС	19042.928
Взаимодействие	Windows Feature Experience Pack 120.2212.551.0

**Reproducibility**
Постоянно в этой ситуации

**Logs/saves**
[vz28pwbe0z1.zip](https://github.com/voxeltycoon/issues/files/6446663/vz28pwbe0z1.zip)
[Player.log](https://github.com/voxeltycoon/issues/files/6446665/Player.log)

**Comment by victorleblais** [comment]
Hello, I don't speak russian, but I translated your message and looked at the attached video.

The issue you have is simply due to an error in signalling: Presignals "relay" the information of signals in front of them. In you case, the presignal is red because there's only one signal in front of it, and that signal is red:
![image](https://user-images.githubusercontent.com/4365075/118292517-003f8d80-b4d9-11eb-9734-b3120ea5aaa0.png)

Then, why does the train want to go here instead of the other side? That's simply because when the train enters a block (by passing a signal) it chooses the destination it wants to go / which block it will try to enter / which signal he will try to pass.

When the train entered the cyan block (in bottom right), the pink block was not occupied, so the presignal between the cyan and blue blocks was green, so the train decided to go there. Then, along the way, the pink block became occupied and thus the signal became red.
Trains don't change their planned route every second (otherwise it would be extremely costly in terms of PC performance).

But don't worry, with proper signalling, this kind of behaviour can be totally avoided. Here's an example of setup:
![image](https://user-images.githubusercontent.com/4365075/118295043-b1472780-b4db-11eb-92bd-e761e469e9d4.png)
And here are some close look:
![image](https://user-images.githubusercontent.com/4365075/118295163-d3d94080-b4db-11eb-8645-45b860d0cfeb.png)
![image](https://user-images.githubusercontent.com/4365075/118295258-ee131e80-b4db-11eb-9d3e-92b6cc8c90a3.png)

Notice the use of presignals and block signals, with those, a train inside a station or a "crossing section" will only enter the shared pieced of tracks if he can safely enter the desired destination or another unoccupied "crossing section".

Also, keep in mind that upgrading to a double-track system will simplify things a lot, and also allow a lot more trains to use your network at once:
![image](https://user-images.githubusercontent.com/4365075/118296151-ed2ebc80-b4dc-11eb-8ce6-7df5d32055b2.png)
![image](https://user-images.githubusercontent.com/4365075/118296194-f9b31500-b4dc-11eb-94e6-d2be3d38b3e8.png)
![image](https://user-images.githubusercontent.com/4365075/118296213-ffa8f600-b4dc-11eb-8071-ea2f453709e5.png)
And if you want to make things even better, consider doing "roll-in roll-out" stations, which prevents train that exit the station from blocking trains entering the station: 
![image](https://user-images.githubusercontent.com/4365075/118296321-1ea78800-b4dd-11eb-9660-7ff71b3acfa9.png)

Notice again the use of presignals and regular block signals.
A good rule to follow if you're not sure is:
- Presignal when entering an intersection
- Block signals when exiting an intersection
